File: typepos.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Iterable, List, cast


@dataclass
class OrderLine:
    unit_price: Decimal
    quantity: int


@dataclass
class NamedOrderLine(OrderLine):
    name: str


def total_price(lines: Iterable[OrderLine]) -> Decimal:
    line_totals = (line.unit_price * line.quantity for line in lines)
    return sum(line_totals, Decimal("0"))


def main():
    line: object = OrderLine(unit_price=Decimal("0.25"), quantity=2)
    order: List[OrderLine] = [
        cast(OrderLine, line),
        OrderLine(unit_price=Decimal("1.00"), quantity=1),
    ]
    print(total_price(order))


if __name__ == "__main__":
    main()

File: typepos.go
package main

import "fmt"

type orderLine struct {
	UnitPrice float64
	Quantity  int
}

type namedOrderLine struct {
	orderLine
	Name string
}

func totalPrice(lines []orderLine) float64 {
	sum := 0.0
	for _, line := range lines {
		sum += line.UnitPrice * float64(line.Quantity)
	}
	return sum
}

func main() {
	var line interface{} = orderLine{UnitPrice: 0.25, Quantity: 2}
	order := []orderLine{
		line.(orderLine),
		{UnitPrice: 1.00, Quantity: 1},
	}
	fmt.Println(totalPrice(order))
}
